File: src/services/MetricsService/MetricsPusher.ts
import {log} from 'shared/logging/config';

export enum MetricsPusherResult {
  Success,
  Error,
}

export interface MetricsPusher {
  push(jsonAsString: string): Promise<MetricsPusherResult>;
}

export class DefaultMetricsPusher implements MetricsPusher {
  private apiEndpointUrl: string;
  private apiEndpointKey: string;

  constructor(apiEndpointUrl: string, apiEndpointKey: string) {
    this.apiEndpointUrl = apiEndpointUrl;
    this.apiEndpointKey = apiEndpointKey;
  }

  push(jsonAsString: string): Promise<MetricsPusherResult> {
    log.debug({
      category: 'metrics',
      message: 'sending metrics to server',
      payload: jsonAsString,
    });
    return fetch(this.apiEndpointUrl, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        'x-api-key': this.apiEndpointKey,
      },
      body: jsonAsString,
    })
      .then(response => response.json())
      .then(json => {
        log.debug({
          category: 'metrics',
          message: 'metrics server response',
          payload: json,
        });
      })
      .then(() => MetricsPusherResult.Success)
      .catch(() => MetricsPusherResult.Error);
  }
}

File: android/app/src/main/java/app/covidshield/services/metrics/MetricsPusher.kt
package app.covidshield.services.metrics

import okhttp3.*

enum class MetricsPusherResult {
    Success, Error
}

interface MetricsPusher {
    fun push(jsonAsString: String): MetricsPusherResult
}

class DefaultMetricsPusher(private val apiEndpointUrl: String, private val apiEndpointKey: String) : MetricsPusher {

    private val okHttpClient by lazy { OkHttpClient() }

    override fun push(jsonAsString: String): MetricsPusherResult {
        val body: RequestBody = RequestBody.create(MediaType.parse("application/json"), jsonAsString)

        val request = Request.Builder()
                .url(apiEndpointUrl)
                .header("Accept", "application/json")
                .addHeader("Content-Type", "application/json")
                .addHeader("x-api-key", apiEndpointKey)
                .cacheControl(CacheControl.Builder().noStore().build())
                .post(body)
                .build()

        return okHttpClient.newCall(request).execute().use { response ->
            if (response.isSuccessful) MetricsPusherResult.Success else MetricsPusherResult.Error
        }
    }

}